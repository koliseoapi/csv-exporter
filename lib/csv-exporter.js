// null, undefined will translate to this value
const EMPTY = "";
const EOL = "\n";

function isNullOrUndefined(object) {
  return typeof object === "undefined" || object === null;
}

// if value is not 100% safe (may contain newline, double quotes, or commas)
// put between quotes and escape any quote
function escape(value) {
  return typeof value !== "string" || !value.length || /^[a-z0-9.@\/+\- ?]+$/i.test(value)
    ? value
    : '"' + value.replace(/"/g, '""') + '"';
}

function createExtractor(property) {
  if (property.indexOf('.') === -1) {
    return function (object) {
      return object[property];
    }
  }
  const parts = property.split('.');
  return function (object) {
    let result = object;
    let i = 0;
    while (!isNullOrUndefined(result) && i < parts.length) {
      result = result[parts[i++]]
    }
    return result;
  }
}

export default function createCsvExporter(config) {
  const labels = [];
  const columns = config.map((configEntry, index) => {
    let property, label;
    if (typeof configEntry === 'string') {
      label = property = configEntry;
    } else {
      property = configEntry.property;
      label = configEntry.label || property;
    }
    if (!property) {
      throw new Error(`Missing property at index ${index}`);
    }
    labels.push(escape(label));
    // function that will retrieve the value of item[property]
    const extractPropertyFrom =
      typeof property === 'function' ? property : createExtractor(property);
    return function (object) {
      if (isNullOrUndefined(object)) return EMPTY;
      const value = extractPropertyFrom(object);
      return isNullOrUndefined(value) ? EMPTY : escape(value);
    };
  });

  return {
    export(data = []) {
      let result = labels.join(",") + EOL;
      data.forEach(object => {
        result += columns.map(callback => callback(object)).join(",") + EOL;
      });
      return result;
    }
  };
}
