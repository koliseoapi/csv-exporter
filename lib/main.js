export function createCsvExporter() {
  return {
    export() {}
  };
}
