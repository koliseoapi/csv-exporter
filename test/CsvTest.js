import { createCsvExporter } from "../lib/main";

describe("Exporter", () => {
  it("renders CSV", () => {
    const csv = createCsvExporter();
    const data = [
      { name: "Bond, James", age: 30, created: new Date("1953-04-13") },
      {
        name: "Sandokan",
        age: 37,
        created: new Date("1895-01-01"),
        description:
          "Emilio Salgari wrote several novels chronicling the adventures of Sandokan and Yanez, two of his most legendary creations.\n The pirates are introduced in The Tigers of Mompracem, which portrays their relentless struggle against the Dutch and British powers that seek to wipe them out. "
      },
      { name: "Mortadelo y Filem√≥n", age: 50.56667 }
    ];
    expect(csv.export(data)).toMatchSnapshot();
  });
});
