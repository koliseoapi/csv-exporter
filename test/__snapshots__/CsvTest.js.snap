// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exporter renders CSV 1`] = `undefined`;
